deepchem/models/torch_model.py

```diff
@@ -0,0 +1,1124 @@
+    """Evaluate the performance of this model on the data produced by a generator.
+
+    Parameters
+    ----------
+    generator: generator
+      this should generate batches, each represented as a tuple of the form
+      (inputs, labels, weights).
+    metric: list of deepchem.metrics.Metric
+      Evaluation metric
+    transformers: list of dc.trans.Transformers
+      Transformers that the input data has been transformed by.  The output
+      is passed through these transformers to undo the transformations.
+    per_task_metrics: bool
+      If True, return per-task scores.
+
+    Returns
+    -------
+    dict
+      Maps tasks to scores under metric.
+    """
+    evaluator = GeneratorEvaluator(self, generator, transformers)
+    return evaluator.compute_model_performance(metrics, per_task_metrics)
+
+  def compute_saliency(self, X: np.ndarray) -> OneOrMany[np.ndarray]:
+    """Compute the saliency map for an input sample.
+
+    This computes the Jacobian matrix with the derivative of each output element
+    with respect to each input element.  More precisely,
+
+    - If this model has a single output, it returns a matrix of shape
+      (output_shape, input_shape) with the derivatives.
+    - If this model has multiple outputs, it returns a list of matrices, one
+      for each output.
+
+    This method cannot be used on models that take multiple inputs.
+
+    Parameters
+    ----------
+    X: ndarray
+      the input data for a single sample
+
+    Returns
+    -------
+    the Jacobian matrix, or a list of matrices
+    """
+    input_shape = X.shape
+    X = np.reshape(X, [1] + list(X.shape))
+    self._ensure_built()
+    X, _, _ = self._prepare_batch(([X], None, None))
+
+    # Compute the gradients.
+
+    X = X[0]
+    X.requires_grad_(True)
+    outputs = self.model(X)
+    if isinstance(outputs, torch.Tensor):
+      outputs = [outputs]
+    final_result = []
+    for output in outputs:
+      output_shape = tuple(output.shape[1:])
+      output = output.reshape([-1])
+      result = []
+      grad_output = torch.zeros(output.shape[0])
+      for i in range(output.shape[0]):
+        grad_output.zero_()
+        grad_output[i] = 1
+        output.backward(grad_output, retain_graph=True)
+        result.append(X.grad.clone())
+        X.grad.zero_()
+      final_result.append(
+          torch.reshape(torch.stack(result),
+                        output_shape + input_shape).numpy())
+    if len(final_result) == 1:
+      return final_result[0]
+    return final_result
+
+  def _prepare_batch(self,
+                     batch: Tuple[Any, Any, Any]) -> Tuple[List, List, List]:
+    inputs, labels, weights = batch
+    inputs = [
+        x.astype(np.float32) if x.dtype == np.float64 else x for x in inputs
+    ]
+    if labels is not None:
+      labels = [
+          x.astype(np.float32) if x.dtype == np.float64 else x for x in labels
+      ]
+      labels = [torch.as_tensor(x) for x in labels]
```

Comment:
I think we should set a value to the device argument.


```python
def get_device():
    device = torch.device("cuda") if torch.cuda.is_available() else torch.device("cpu")
    return device

device = get_device()
torch.as_tensor(a, device=device)
```

Reply:
Good suggestion.  I haven't tested any of this on a GPU yet.  That's next up on my list.